### ISSUE TITLE
Publish a mechanism for disabling subclass preservation of Tensor subclasses

### ISSUE THREAD

**Issue by ferrine** [pytorch/pytorch#46159]
## 🐛 Bug

In inherited `torch.Tensor` ambiguous type promotions appear (nightly release).

## To Reproduce

```python
class MyTensor(torch.Tensor):
    ...
ten = MyTensor(10)
assert isinstance(ten, MyTensor)
assert isinstance(ten + 1, MyTensor), "I expect this to fail"
```
## Expected behavior
```python
class MyTensor(torch.Tensor):
    ...
ten = MyTensor(10)
assert isinstance(ten, MyTensor)
assert not isinstance(ten + 1, MyTensor), "all good"
```
Or a way to properly subclass from `torch.Tensor` avoiding such issues

## Environment

```
PyTorch version: 1.8.0.dev20201011+cpu
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: 10.0.1 
CMake version: version 3.18.4

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: 10.0.130
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
```

## Additional context

I use subclassing in [geoopt](https://github.com/geoopt/geoopt) and maintain the project

cc @nairbv @mruberry

**Comment by glaringlee** [comment]
@ferrine
You can override __add__ in your class, do type checking and your own impl for addition operation.

**Comment by ferrine** [comment]
I'm sure it is not a maintainable way. I would override every single class method because I want pytorch behavior like in old releases. It just performed regular addition and returned pytorch tensors

**Comment by mruberry** [comment]
@ezyang This seems related to work like torch_function? Is there a good label for these issues? I don't think it's type promotion.

**Comment by ezyang** [comment]
cc @hameerabbasi

**Comment by ezyang** [comment]
Yes, this is a behavior change introduced by `__torch_function__`. Hameer and I discussed in the PR whether or not we should BC break or not (https://github.com/pytorch/pytorch/pull/37091#pullrequestreview-419541245). Fortunately, there is a way to recover the old behavior:

```
from torch._C import _disabled_torch_function_impl

class MyTensor(torch.Tensor):
    __torch_function__ = _disabled_torch_function_impl
```

I was checking the release notes and I realized we forgot to add a note about this. Part of the issue is we didn't actually publish this API (https://github.com/pytorch/pytorch/pull/37091#issuecomment-658743122) but I think @rgommers et al agreed that we could add a public API later. For now, you should use the internal API, I think.

**Comment by hameerabbasi** [comment]
Perhaps an alias of `torch._C._disabled_torch_function_impl` in `torch.overrides` would work?

**Comment by rgommers** [comment]
> Or a way to properly subclass from `torch.Tensor` avoiding such issues

https://pytorch.org/docs/master/notes/extending.html#subclassing-torch-tensor is new, subclassing `Tensor` should now be much more "proper", would be good to have a look if that fits your needs.

> Perhaps an alias of `torch._C._disabled_torch_function_impl` in `torch.overrides` would work?

I'd still prefer people to first use the private API as a workaround and see if there's actual use cases before making this public. This issue does seem to suggest better subclassing behaviour is the preferred option.

**Comment by ezyang** [comment]
Looking at https://github.com/geoopt/geoopt it looks like the use case here is implementing ManifoldTensor and ManifoldParameter. I'm guessing that they want to discard manifold implementation when a non-manifold-aware operation is used.

Definitions are at https://github.com/geoopt/geoopt/blob/master/geoopt/tensor.py

**Comment by ferrine** [comment]
> Looking at https://github.com/geoopt/geoopt it looks like the use case here is implementing ManifoldTensor and ManifoldParameter. I'm guessing that they want to discard manifold information when a non-manifold-aware operation is used.
> 
> Definitions are at https://github.com/geoopt/geoopt/blob/master/geoopt/tensor.py

Yes, I also had very long discussions about whether should I make any operation manifold-aware or not. Given the complications I faced in prior releases I decided not to experiment with exposing my prototypes and just keep the rest simple.

**Comment by rgommers** [comment]
I played around with `ManifoldTensor` a bit. Given that there are (AFAICT) no examples in the docs it's a little hard to understand the exact goals and I may be missing something important, but it seems to me that the API of `ManifoldTensor` is a little confusing.

When you construct a `ManifoldTensor`, it has a `.manifold` attribute (the extra "metadata"), all the regular methods that `torch.Tensor` also has, and some extra methods. 
```
>>> torch.__version__                                                      
'1.5.0'
>>> t = geoopt.ManifoldTensor([1,2,3], manifold=geoopt.Euclidean())
>>> t                                                                      
Tensor on Euclidean(ndim=0) manifold containing:
tensor([1., 2., 3.])
>>> t[:2]                                                                  
tensor([1., 2.])
>>> t + 1                                                                  
tensor([2., 3., 4.])
>>> t * t                                                                  
tensor([1., 4., 9.])
>>> t.proju(t)                                                             
tensor([1., 2., 3.])
>>> t.transp(t, t)                                                         
tensor([1., 2., 3.])
>>> t.proj_()                                                              
Tensor on Euclidean(ndim=0) manifold containing:
tensor([1., 2., 3.])
```
Using pretty much any method except `proj_` loses the subclass. So it's a "construct, use once, then treat as regular Tensor" kind of construct. I understand that now that you have code using this you want to keep it, but I'm unclear on why this is a design pattern we may want to support long-term. If operations like slicing, or adding two instances together, or calling `.reshape(), `.max()` or other common methods all lose the extra metadata you tacked onto a tensor, they become semantically different things that you'd probably not want users to use.

If I'd have to implement `ManifoldTensor`, I think I'd make it a regular class (not a subclass) with `.manifold` and `.tensor` attributes and _only_ the few methods that are manifold-aware. And make the user explicitly drop the subclass if that's what they wanted (that's a one-liner).

> Yes, I also had very long discussions about whether should I make any operation manifold-aware or not. Given the complications I faced in prior releases I decided not to experiment with exposing my prototypes and just keep the rest simple.

That was the main point of this bc-breaking change: it used to be very complicated, and now it should be doing "the right thing" without you having to reimplement any of the standard methods. This is also how NumPy subclasses work, and I don't think I have ever seen a request or good use case for "we silently want to drop our subclass here".

**Comment by rgommers** [comment]
> it is recognized as a parameter and recorded by `torch.nn.Module`, same for `geoopt.ManifoldTensor` that can be a buffer.

Thanks, that helps. Then the alternative would be:
```
def manifold_tensor(tensor, manifold=Euclidean()):
    # do some manifold validation here
    tensor.manifold = manifold
    return tensor
```
All the methods are actually defined on the manifold, e.g.:
```
    def retr(self, ...):
        return self.manifold.retr(...)
```
and (modulo a tweak to `__repr__` and pickling support) it seems everything interesting is in `.manifold`.

I'm not suggesting you actually make a change (https://github.com/pytorch/pytorch/issues/46159#issuecomment-707817037 should fix your issue right now), but I think you semantically don't really want a subclass, but rather simply a `torch.Tensor` (and `Parameter`) with the ability to carry around a single extra attribute (`.manifold`). You're now actually forced to use private API (e.g., `torch._utils._rebuild_tensor_v2`; a (slightly out of sync) duplicate of `__reduce_ex__`) that may be subject to change to achieve that.

**Comment by ferrine** [comment]
> Yes, this is a behavior change introduced by `__torch_function__`. Hameer and I discussed in the PR whether or not we should BC break or not ([#37091 (review)](https://github.com/pytorch/pytorch/pull/37091#pullrequestreview-419541245)). Fortunately, there is a way to recover the old behavior:
> 
> ```
> from torch._C import _disabled_torch_function_impl
> 
> class MyTensor(torch.Tensor):
>     __torch_function__ = _disabled_torch_function_impl
> ```
> 
> I was checking the release notes and I realized we forgot to add a note about this. Part of the issue is we didn't actually publish this API ([#37091 (comment)](https://github.com/pytorch/pytorch/pull/37091#issuecomment-658743122)) but I think @rgommers et al agreed that we could add a public API later. For now, you should use the internal API, I think.

I played around with different suggestions, this was the best one
